aiohttp_s3_client/credentials.py

```diff
@@ -0,0 +1,158 @@
+import json
+import os
+import socket
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Union
+from urllib.request import urlopen
+
+from yarl import URL
+
+
+@dataclass(frozen=True)
+class Credentials:
+    access_key_id: str
+    secret_access_key: str
+    region: str = ""
+    service: str = "s3"
+
+    def __bool__(self) -> bool:
+        return all((self.access_key_id, self.secret_access_key))
+
+    def __repr__(self) -> str:
+        return (
+            f"{self.__class__.__name__}(access_key_id={self.access_key_id!r}, "
+            "secret_access_key="
+            f'{"******" if self.secret_access_key else None!r}, '
+            f"region={self.region!r}, service={self.service!r})"
+        )
+
+    def as_dict(self) -> dict:
+        return {
+            "region": self.region,
+            "access_key_id": self.access_key_id,
+            "secret_access_key": self.secret_access_key,
+            "service": self.service,
+        }
+
+
+class URLCredentials(Credentials):
+    def __init__(
+        self, url: Union[str, URL], *, region: str = "", service: str = "s3"
+    ):
+        url = URL(url)
+        super().__init__(
+            access_key_id=url.user or '',
+            secret_access_key=url.password or '',
+            region=region, service=service,
+        )
+
+
+class EnvironmentCredentials(Credentials):
+    def __init__(self, region: str = '', service: str = "s3"):
+        super().__init__(
+            access_key_id=os.getenv("AWS_ACCESS_KEY_ID", ''),
+            secret_access_key=(
+                os.getenv("AWS_SECRET_ACCESS_KEY", '') or
+                os.getenv("AWS_SESSION_TOKEN", '')
+            ),
+            region=os.getenv('AWS_DEFAULT_REGION', region),
+            service=service,
+        )
+
+
+class ConfigCredentials(Credentials):
+    DEFAULT_CONFIG_PATH = Path.home() / '.aws' / 'credentials'
+
+    def __init__(
+        self, config_path: Union[str, Path] = DEFAULT_CONFIG_PATH, *,
+        region: str = '', service: str = "s3", profile: str = 'auto'
+    ):
+        credentials_file = os.getenv('AWS_SHARED_CREDENTIALS_FILE')
+        if credentials_file is not None:
+            config_path = Path(credentials_file)
+
+        if not isinstance(config_path, Path):
+            config_path = Path(config_path)
+
+        if not config_path.exists():
+            super().__init__(
+                access_key_id='', secret_access_key='',
+                region=region, service=service,
+            )
+            return
+
+        if profile == 'auto':
+            profile = os.getenv('AWS_PROFILE', 'default')
+
+        creds = configparser.ConfigParser()
+        creds.read(config_path)
+
+        if profile not in creds:
+            super().__init__(
+                access_key_id='', secret_access_key='',
+                region=region, service=service,
+            )
+            return
+
+        section = creds[profile]
+
+        super().__init__(
```

Comment:
there are two "config" files, creds and config.  the utility "aws configure" will write region into ~/.aws/config and accesskey/secretkey into ~/.aws/credentials.  

as per https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html

what's annoying, that according to docs you can pick up creds from config, but it doesn't seem that it's possible the other way around (e.g. region in creds).  regardless, i'd expect the normal setup to be the results of "aws configure" which is as I stated region in config and the rest in credentials

also, if reading from config, the location of config can be overridden by AWS_CONFIG_FILE env var.

Reply:
done